### PR TITLE
CAMEL-9638: Introduced SNIHostName config in sslContextParameters

### PR DESCRIPTION
**Pull request by anoordover**
Introduced SNIHostName in sslContextParameters.
Added a test to https://www.mnot.net which throws 403 or 421 when SNIHostName is not used.
Eventually it might be needed to ignore this test. I didn't find a simple way to test this.

### CODE REVIEW

**Comment by anoordover** [comment]
@davsclaus and @oscerd can you review this?
Can you advice on the test shall I ignore it in the next commit?

File: components/camel-core-xml/src/main/java/org/apache/camel/core/xml/util/jsse/SNIHostNamesDefinition.java [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+
+import java.util.ArrayList;
+import java.util.List;
+import javax.xml.bind.annotation.XmlAccessType;
+import javax.xml.bind.annotation.XmlAccessorType;
+import javax.xml.bind.annotation.XmlElement;
+import javax.xml.bind.annotation.XmlType;
+
+@XmlAccessorType(XmlAccessType.FIELD)
+@XmlType(name = "SNIHostNames", propOrder = {"sniHostName"})
```
Comment (davsclaus) [inline components/camel-core-xml/src/main/java/org/apache/camel/core/xml/util/jsse/SNIHostNamesDefinition.java]: All the types in the Camel DSL are lower case as start, eg should be sniHostNames instead.

File: components/camel-http4/src/test/resources/org/apache/camel/component/http4/CamelHttp4Context.xml [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+       xsi:schemaLocation="
+       http://www.springframework.org/schema/beans http://www.springframework.org/schema/beans/spring-beans.xsd
+       http://camel.apache.org/schema/spring http://camel.apache.org/schema/spring/camel-spring.xsd
+    ">
+
+  <bean class="org.springframework.beans.factory.config.PropertyPlaceholderConfigurer"/>
+
+  <camel:sslContextParameters id="correctSniSslContextParameters">
+    <camel:clientParameters>
+      <camel:SNIHostNamesDefinition>
```
Comment (davsclaus) [inline components/camel-http4/src/test/resources/org/apache/camel/component/http4/CamelHttp4Context.xml]: We should not use the name xxxDefiniton in the DSL that the end user sees. It should just be sniHostNames.

File: camel-core/src/main/java/org/apache/camel/util/jsse/SSLContextClientParameters.java [hunk 3]
```diff
@@ -34,6 +37,19 @@
     
     private static final Logger LOG = LoggerFactory.getLogger(SSLContextClientParameters.class);
 
+    private List<SNIServerName> sniHostNames = new ArrayList<>();
+
+    public void setSniHostNames(List<String> sniHostNames) {
```
Comment (davsclaus) [inline camel-core/src/main/java/org/apache/camel/util/jsse/SSLContextClientParameters.java]: Maybe let it be a plain getter/setter. As this setter will add and not remove any existing. A bit unusual, as getter/setter are usually just to set or get the values as-is.

File: camel-core/src/main/java/org/apache/camel/util/jsse/BaseSSLContextParameters.java [hunk 4]
```diff
@@ -515,7 +521,11 @@ public SSLServerSocketFactory configure(SSLServerSocketFactory factory) {
             
             @Override
             public SSLSocket configure(SSLSocket socket) {
-                
+
+                SSLParameters sslParameters = socket.getSSLParameters();
+                sslParameters.setServerNames(getSNIHostNames());
```
Comment (davsclaus) [inline camel-core/src/main/java/org/apache/camel/util/jsse/BaseSSLContextParameters.java]: I wonder if you should only call this setServiceNames if the host names are not empty (eg there is some values)

**Comment by anoordover** [comment]
@davsclaus, fixed all your review notes. Thanks.
Is it ready to be pulled?
Or shall I ignore the test that does a get on www.mnot.net?

**Comment by davsclaus** [comment]
Sounds good. And yeah ignore any tests that rely on online internet connectivity.

**Pull request merged**